wallet/wallet.go

```diff
@@ -230,6 +237,107 @@ func (w *SingleAddressWallet) SignTransaction(cs consensus.State, txn *types.Tra
 	return nil
 }
 
+// Split returns a transaction that splits the wallet in the given number of
+// outputs with given amount. The transaction is funded but not signed.
+//
+// NOTE: split needs to use a minimal set of inputs and therefore does not reuse
+// the fund logic which randomizes the unspent transaction outputs used to fund
+// the transaction
+func (w *SingleAddressWallet) Split(cs consensus.State, outputs int, amount, feePerByte types.Currency, pool []types.Transaction) (types.Transaction, []types.OutputID, error) {
+	// prepare all outputs
+	var txn types.Transaction
+	for i := 0; i < int(outputs); i++ {
+		txn.SiacoinOutputs = append(txn.SiacoinOutputs, types.SiacoinOutput{
+			Value:      amount,
+			UnlockHash: w.Address(),
+		})
+	}
+
+	// fetch unspent transaction outputs
+	utxos, err := w.store.UnspentSiacoinElements()
+	if err != nil {
+		return types.Transaction{}, nil, err
+	}
+
+	// desc sort
+	sort.Slice(utxos, func(i, j int) bool {
+		return utxos[i].Value.Cmp(utxos[j].Value) > 0
+	})
+
+	// map used outputs
+	inPool := make(map[types.OutputID]bool)
+	for _, ptxn := range pool {
+		for _, in := range ptxn.SiacoinInputs {
+			inPool[types.OutputID(in.ParentID)] = true
+		}
+	}
+
+	// filter outputs that are in use
+	var u int
+	for _, sce := range utxos {
+		inUse := w.used[sce.ID] || inPool[sce.ID]
+		matured := cs.Index.Height >= sce.MaturityHeight
+		sameValue := sce.Value.Equals(amount)
+		if !inUse && !sameValue && matured {
+			utxos[u] = sce
+			u++
+		}
+	}
+	utxos = utxos[:u]
+
+	// estimate the fees
+	outputFees := feePerByte.Mul64(uint64(len(encoding.Marshal(txn.SiacoinOutputs))))
+	feePerInput := feePerByte.Mul64(BytesPerInput)
+
+	// search for minimal set
+	want := amount.Mul64(uint64(outputs))
+	i := sort.Search(len(utxos)+1, func(i int) bool {
+		fee := feePerInput.Mul64(uint64(i)).Add(outputFees)
+		return SumOutputs(utxos[:i]).Cmp(want.Add(fee)) >= 0
+	})
```

Comment:
I think I was being a bit too clever when I wrote this. A simple `for` loop would work fine. (Plus, then we could perform the filtering step within the same loop.)